odltools/karaf/dump.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2018 Red Hat, Inc. and others. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+import os
+
+line_num = 0
+OPENER = '[{<'
+CLOSER = ']}>'
+INDENT = '  '
+SKIP_FIELDS = 4
+
+
+def dump_karaf_log(args):
+    path = args.path
+    if path == '-':
+        _dump_pretty_print(sys.stdin)
+        return
+
+    if not os.path.isfile(path):
+        sys.stderr.write('File "%s" not found' % path)
+        return
+
+    with open(path, 'r') as log_file:
+        _dump_pretty_print(log_file)
+
+
+def _has_nested_structs(msg):
+    top = -1
+    for c in msg:
+        if top == -1:
+            top = OPENER.find(c)
+        else:
+            if c in OPENER:
+                return True
+            closer = CLOSER.find(c)
+            if closer == top:
+                top = -1
+    return False
+
+
+def _nl(indent):
+    global line_num
+    line_num += 1
+    sys.stdout.write('\n' + indent)
+
+
+def _dump_pretty_print(input):
+    for line in input:
```

Comment:
one suggestion here is instead of keeping track of indenting and printing line by line, you could just store all of this as json object and then do a json.dumps(indent=x) on a per parsed line basis.  Don't have to do it but it might be more simple and will pretty print it.

Reply:
The dumped format is not valid json and I wanted to avoid transforming it to such and keep it identical to the original except for new lines and white space.

Reply:
yeah, I was going to post the same idea to dump as json, but then I recalled this is from a toString() or similar so it isn't valid json.

Reply:
I actually spoke to Robert Varga about perhaps changing that toString to give valid json since the output is SO close. He objected.